google/api_core/retry/retry_base.py

```diff
@@ -318,26 +309,42 @@ def with_predicate(self, predicate: Callable[[Exception], bool]) -> Self:
         Returns:
             Retry: A new retry instance with the given predicate.
         """
-        return self._replace(predicate=predicate)
+        return type(self)(
+            predicate=predicate,
+            initial=self._initial,
+            maximum=self._maximum,
+            multiplier=self._multiplier,
+            timeout=self._timeout,
+            on_error=self._on_error,
+        )
 
     def with_delay(
         self,
-        initial: float | None = None,
-        maximum: float | None = None,
-        multiplier: float | None = None,
+        initial: Optional[float] = None,
+        maximum: Optional[float] = None,
+        multiplier: Optional[float] = None,
     ) -> Self:
         """Return a copy of this retry with the given delay options.
 
         Args:
             initial (float): The minimum amount of time to delay (in seconds). This must
-                be greater than 0.
-            maximum (float): The maximum amount of time to delay (in seconds).
-            multiplier (float): The multiplier applied to the delay.
+                be greater than 0. If None, the current value is used.
+            maximum (float): The maximum amount of time to delay (in seconds). If None, the
+                current value is used.
+            multiplier (float): The multiplier applied to the delay. If None, the current
+                value is used.
 
         Returns:
-            Retry: A new retry instance with the given predicate.
+            Retry: A new retry instance with the given delay options.
         """
-        return self._replace(initial=initial, maximum=maximum, multiplier=multiplier)
+        return type(self)(
+            predicate=self._predicate,
+            initial=initial if initial is not None else self._initial,
```

Comment:
If these values are `not None` but still falsy, we still want to use them? If so, we should capture this in a test since this differs from the previous behavior.

Reply:
Yeah it differs from the previous behaviour, but it is in line with the [pre-2.16.0 behaviour](https://github.com/googleapis/python-api-core/blob/9866183b42f632dddade9ba6c69f9b3015118dd4/google/api_core/retry.py#L457)

Good point on the test. I had added one for timeout, but not for this one. Fixed

Reply:
@daniel-sanche  Please could you link the specific commit for the fix in https://github.com/googleapis/python-api-core/pull/592#discussion_r1476827410?

Reply:
sorry, forgot to push. The new test should be there now